jwst/extract_2d/grisms.py

```diff
@@ -304,68 +318,89 @@ def extract_grism_objects(input_model,
+            # row or column of the detector
+            if (((ymax - ymin) > 0) and ((xmax - xmin) > 0)):
+                # only the first two numbers in the Mapping are used
+                # the order and source position are put directly into
+                # the new wcs for the subarray for the forward transform
+                xcenter_model = Const1D(obj.xcentroid)
+                xcenter_model.inverse = Const1D(obj.xcentroid)
+                ycenter_model = Const1D(obj.ycentroid)
+                ycenter_model.inverse = Const1D(obj.ycentroid)
+                order_model = Const1D(order)
+                order_model.inverse = Const1D(order)
+                tr = inwcs.get_transform('grism_detector', 'detector')
+                tr = Identity(2) | Mapping((0, 1, 0, 1, 0)) | (Shift(xmin) & Shift(ymin) &
+                                                               xcenter_model &
+                                                               ycenter_model &
+                                                               order_model) | tr
+
+                subwcs.set_transform('grism_detector', 'detector', tr)
+
+                log.info("Subarray extracted for obj: {} order: {}:".format(obj.sid, order))
+                log.info("Subarray extents are: (xmin:{}, xmax:{}), (ymin:{}, ymax:{})".format(xmin, xmax, ymin, ymax))
+
+                ext_data = input_model.data[ymin: ymax + 1, xmin: xmax + 1].copy()
+                ext_err = input_model.err[ymin: ymax + 1, xmin: xmax + 1].copy()
+                ext_dq = input_model.dq[ymin: ymax + 1, xmin: xmax + 1].copy()
+
+                # new_model = datamodels.ImageModel(data=ext_data, err=ext_err, dq=ext_dq)
+                new_model = datamodels.SlitModel(data=ext_data, err=ext_err, dq=ext_dq)
+                new_model.meta.wcs = subwcs
+                new_model.meta.wcsinfo.spectral_order = order
+
+                # set x/ystart values relative to the image (screen) frame.
+                # The overall subarray offset is recorded in model.meta.subarray.
+                # nslit = obj.sid - 1  # catalog id starts at zero
+                log.debug("sid {} xcen {}  ycen {}".format(obj.sid, obj.xcentroid, obj.ycentroid))
+                new_model.name = str(obj.sid)
+                new_model.xstart = xmin + 1
+                new_model.xsize = (xmax - xmin)
+                new_model.ystart = ymin + 1
+                new_model.ysize = (ymax - ymin)
+                new_model.source_xpos = float(obj.xcentroid)
+                new_model.source_ypos = float(obj.ycentroid)
+                new_model.source_id = obj.sid
+                new_model.bunit_data = input_model.meta.bunit_data
+                new_model.bunit_err = input_model.meta.bunit_err
+                slits.append(new_model)
+
     output_model.slits.extend(slits)
     del subwcs
     return output_model
 
 
+def clamp(value, minval, maxval):
```

Comment:
yah, wasn't sure where to stick it so was waiting for comment. There's no general use util in lib/  but we could make one

Reply:
astropy.stats would also be a good place -- this comes up a lot in astronomy.    I think it is okay to leave it here for now.

Reply:
Why not use `np.clip`, which works on scalars or arrays?

Reply:
hmm, maybe? I just wrote that function recently to clean up how the code read, lemme go look at the specifics of clip again

Reply:
...maybe cause it's always making arrays?

Reply:
That would be my guess.  Unfortunately there isn't a `math.clip` function.

How many times are you calling this function such that a difference of < 4 µs is noticeable? :smile:

Reply:
every object in randomly sized catalogs plus their corners?

Reply:
if agreeable to everyone, since numpy isn't used elsewhere in this module, and the clamp call is slightly faster, I think I'm going to keep it...we can always change it later!

Reply:
I'm okay with that.